File: src/python/catalog.py
"""Executable, dependency-free contracts for the complete pattern manifest."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from threading import Lock
from typing import Callable

EXPECTED_PATTERN_COUNT = 253


@dataclass(frozen=True)
class PatternDefinition:
    identifier: str
    catalogues: tuple[str, ...]
    family: str
    name: str
    contract: str


def parse_catalog(path: Path) -> list[PatternDefinition]:
    definitions: list[PatternDefinition] = []
    for line_number, raw_line in enumerate(path.read_text(encoding="utf-8").splitlines(), 1):
        if not raw_line or raw_line.startswith("#"):
            continue
        fields = raw_line.split("|")
        if len(fields) != 5:
            raise ValueError(f"Malformed manifest record on line {line_number}")
        identifier, catalogues, family, name, contract = fields
        definitions.append(PatternDefinition(identifier, tuple(catalogues.split(";")), family, name, contract))
    return definitions


def contracts() -> dict[str, Callable[[], bool]]:
    def blackboard() -> bool:
        shared_facts: list[str] = []

        def extract_tokens() -> None:
            shared_facts.extend(["candidate:invoice", "amount:42"])

        def infer_classification() -> None:
            if "candidate:invoice" in shared_facts:
                shared_facts.append("classification:billable")

        extract_tokens()
        infer_classification()
        return shared_facts[-1] == "classification:billable"

    def broker() -> bool:
        service_registry = {"pricing": lambda product_id: f"quote:{product_id}"}

        def request(service_name: str, product_id: str) -> str:
            return service_registry[service_name](product_id)

        return request("pricing", "42") == "quote:42"

    def pac() -> bool:
        abstraction = {"selected": "none"}

        def control(action: str) -> None:
            abstraction["selected"] = action.removeprefix("select:")

        def presentation() -> str:
            return f"selected:{abstraction['selected']}"

        control("select:report")
        return presentation() == "selected:report"

    def reflection() -> bool:
        class Formatter:
            @staticmethod
            def upper(value: str) -> str:
                return value.upper()

        operation = getattr(Formatter, "upper")
        return operation("catalogue") == "CATALOGUE"

    def master_slave() -> bool:
        workers = [lambda value: value * 2, lambda value: value * 3]
        results = [worker(2) for worker in workers]
        return sum(results) == 10

    def command_processor() -> bool:
        executed: list[str] = []
        pending = [lambda: executed.append("refresh")]
        pending.pop(0)()
        return executed == ["refresh"]

    def view_handler() -> bool:
        commands = {"save-clicked": lambda: "save-document"}
        return commands["save-clicked"]() == "save-document"

    def forwarder_receiver() -> bool:
        def receiver(payload: str) -> str:
            return f"received:{payload}"

        def forwarder(message: str) -> str:
            return receiver(message.removeprefix("send:"))

        return forwarder("send:invoice") == "received:invoice"

    def whole_part() -> bool:
        class Folder:
            def __init__(self) -> None:
                self._parts: list[int] = []

            def add_part(self, size: int) -> None:
                self._parts.append(size)

            def total_size(self) -> int:
                return sum(self._parts)

        archive = Folder()
        archive.add_part(3)
        archive.add_part(5)
        return archive.total_size() == 8

    def client_dispatcher_server() -> bool:
        servers = {"calculate": lambda payload: f"result:{payload * 2}"}

        def dispatcher(operation: str, payload: int) -> str:
            return servers[operation](payload)

        def client() -> str:
            return dispatcher("calculate", 21)

        return client() == "result:42"

    def counted_pointer() -> bool:
        class SharedHandle:
            def __init__(self, value: str) -> None:
                self.value = value
                self.references = 1

            def acquire(self) -> "SharedHandle":
                self.references += 1
                return self

            def release(self) -> int:
                self.references -= 1
                return self.references

        document = SharedHandle("invoice")
        observer = document.acquire()
        return observer.value == "invoice" and document.release() == 1

    def wrapper_facade() -> bool:
        class SocketLibrary:
            def connect(self, host: str, port: int) -> str:
                return f"{host}:{port}"

        class CatalogueConnection:
            def __init__(self, socket_library: SocketLibrary) -> None:
                self._socket_library = socket_library

            def open(self) -> str:
                return self._socket_library.connect("catalogue", 443)

        return CatalogueConnection(SocketLibrary()).open() == "catalogue:443"

    def component_configurator() -> bool:
        class Cache:
            def start(self) -> str:
                return "cache:ready"

        class ComponentConfigurator:
            def __init__(self, factories: dict[str, Callable[[], Cache]]) -> None:
                self._factories = factories

            def configure(self, name: str) -> str:
                return self._factories[name]().start()

        return ComponentConfigurator({"cache": Cache}).configure("cache") == "cache:ready"

    def interceptor() -> bool:
        def audit_interceptor(request: dict[str, str], next_handler: Callable[[dict[str, str]], str]) -> str:
            request["audit"] = "recorded"
            return next_handler(request)

        return audit_interceptor({"operation": "save"}, lambda request: request["audit"]) == "recorded"

    def extension_interface() -> bool:
        class Diagnostics:
            def status(self) -> str:
                return "diagnostics:ready"

        class CatalogueComponent:
            def extension(self, name: str) -> Diagnostics | None:
                return Diagnostics() if name == "diagnostics" else None

        diagnostics = CatalogueComponent().extension("diagnostics")
        return diagnostics is not None and diagnostics.status() == "diagnostics:ready"

    def asynchronous_completion_token() -> bool:
        class CompletionToken:
            def __init__(self, request_id: str) -> None:
                self.request_id = request_id
                self.result: str | None = None

            def complete(self, result: str) -> None:
                self.result = result

        token = CompletionToken("run-42")
        token.complete("saved")
        return token.request_id == "run-42" and token.result == "saved"

    def acceptor_connector() -> bool:
        class Acceptor:
            def accept(self, peer: str) -> str:
                return f"connected:{peer}"

        class Connector:
            def __init__(self, acceptor: Acceptor) -> None:
                self._acceptor = acceptor

            def connect(self, peer: str) -> str:
                return self._acceptor.accept(peer)

        return Connector(Acceptor()).connect("catalogue-client") == "connected:catalogue-client"

    def scoped_locking() -> bool:
        lock = Lock()
        protected_state: list[str] = []

        with lock:
            protected_state.append("updated")

        return protected_state == ["updated"] and not lock.locked()

    return {
        "boundary": lambda: {"request": "accepted"}["request"] == "accepted",
        "blackboard": blackboard,
        "broker": broker,
        "pac": pac,
        "reflection": reflection,
        "master-slave": master_slave,
        "command-processor": command_processor,
        "view-handler": view_handler,
        "forwarder-receiver": forwarder_receiver,
        "whole-part": whole_part,
        "client-dispatcher-server": client_dispatcher_server,
        "counted-pointer": counted_pointer,
        "wrapper-facade": wrapper_facade,
        "component-configurator": component_configurator,
        "interceptor": interceptor,
        "extension-interface": extension_interface,
        "asynchronous-completion-token": asynchronous_completion_token,
        "acceptor-connector": acceptor_connector,
        "scoped-locking": scoped_locking,
        "composition": lambda: "|".join(["first", "second"]) == "first|second",
        "concurrency": lambda: len({"leader"}) == 1,
        "deployment": lambda: {"region-a", "region-b"} == {"region-a", "region-b"},
        "mapping": lambda: {"external": "internal"}["external"] == "internal",
        "messaging": lambda: (lambda message: message["id"] == "m-1")({"id": "m-1"}),
        "observability": lambda: all(["trace-1", "healthy"]),
        "ordering": lambda: sorted([3, 1, 2]) == [1, 2, 3],
        "persistence": lambda: (lambda store: store.get("id") == "saved")({"id": "saved"}),
        "resilience": lambda: next(attempt for attempt in range(1, 4) if attempt == 2) == 2,
        "routing": lambda: ("billing" if "invoice".startswith("invoice") else "support") == "billing",
        "security": lambda: {"token": "scoped"}["token"] == "scoped",
        "state": lambda: (lambda state: state == "ready")("ready"),
    }


def verify(definitions: list[PatternDefinition]) -> None:
    available_contracts = contracts()
    if len(definitions) != EXPECTED_PATTERN_COUNT:
        raise AssertionError(f"Expected {EXPECTED_PATTERN_COUNT} records, found {len(definitions)}")
    identifiers = [definition.identifier for definition in definitions]
    if len(set(identifiers)) != len(identifiers):
        raise AssertionError("Pattern identifiers must be unique")
    for definition in definitions:
        if not all([definition.identifier, definition.catalogues, definition.family, definition.name]):
            raise AssertionError(f"Incomplete definition: {definition.identifier}")
        contract = available_contracts.get(definition.contract)
        if contract is None or not contract():
            raise AssertionError(f"{definition.name} contract failed")
        print(f"PASS {definition.name}")


def main() -> None:
    manifest = Path(__file__).resolve().parents[1] / "shared" / "pattern-catalog.tsv"
    verify(parse_catalog(manifest))
    print(f"Verified {EXPECTED_PATTERN_COUNT} catalogued patterns.")


if __name__ == "__main__":
    main()

File: src/javascript/catalog.js
"use strict";

const { readFileSync } = require("node:fs");
const { join } = require("node:path");

const EXPECTED_PATTERN_COUNT = 259;

function blackboardContract() {
  const sharedFacts = [];
  const extractTokens = () => sharedFacts.push("candidate:invoice", "amount:42");
  const inferClassification = () => {
    if (sharedFacts.includes("candidate:invoice")) sharedFacts.push("classification:billable");
  };
  extractTokens();
  inferClassification();
  return sharedFacts.at(-1) === "classification:billable";
}

function brokerContract() {
  const serviceRegistry = new Map([["pricing", (productId) => `quote:${productId}`]]);
  const request = (serviceName, productId) => serviceRegistry.get(serviceName)(productId);
  return request("pricing", "42") === "quote:42";
}

function pacContract() {
  const abstraction = { selected: "none" };
  const control = (action) => { abstraction.selected = action.replace("select:", ""); };
  const presentation = () => `selected:${abstraction.selected}`;
  control("select:report");
  return presentation() === "selected:report";
}

function reflectionContract() {
  const formatter = { upper: (value) => value.toUpperCase() };
  const operation = formatter["upper"];
  return operation("catalogue") === "CATALOGUE";
}

function masterSlaveContract() {
  const workers = [(value) => value * 2, (value) => value * 3];
  const results = workers.map((worker) => worker(2));
  return results.reduce((sum, result) => sum + result, 0) === 10;
}

function commandProcessorContract() {
  const executed = [];
  const pending = [() => executed.push("refresh")];
  pending.shift()();
  return executed.join(",") === "refresh";
}

function viewHandlerContract() {
  const commands = new Map([["save-clicked", () => "save-document"]]);
  return commands.get("save-clicked")() === "save-document";
}

function forwarderReceiverContract() {
  const receiver = (payload) => `received:${payload}`;
  const forwarder = (message) => receiver(message.replace("send:", ""));
  return forwarder("send:invoice") === "received:invoice";
}

function wholePartContract() {
  class Folder {
    #parts = [];
    addPart(size) { this.#parts.push(size); }
    totalSize() { return this.#parts.reduce((total, size) => total + size, 0); }
  }
  const archive = new Folder();
  archive.addPart(3);
  archive.addPart(5);
  return archive.totalSize() === 8;
}

function clientDispatcherServerContract() {
  const servers = new Map([["calculate", (payload) => `result:${payload * 2}`]]);
  const dispatcher = (operation, payload) => servers.get(operation)(payload);
  const client = () => dispatcher("calculate", 21);
  return client() === "result:42";
}

function countedPointerContract() {
  class SharedHandle {
    #references = 1;
    constructor(value) { this.value = value; }
    acquire() { this.#references += 1; return this; }
    release() { this.#references -= 1; return this.#references; }
  }

  const document = new SharedHandle("invoice");
  const observer = document.acquire();
  return observer.value === "invoice" && document.release() === 1;
}

function wrapperFacadeContract() {
  class SocketLibrary { connect(host, port) { return `${host}:${port}`; } }
  class CatalogueConnection {
    constructor(socketLibrary) { this.socketLibrary = socketLibrary; }
    open() { return this.socketLibrary.connect("catalogue", 443); }
  }
  return new CatalogueConnection(new SocketLibrary()).open() === "catalogue:443";
}

function componentConfiguratorContract() {
  class Cache { start() { return "cache:ready"; } }
  class ComponentConfigurator {
    constructor(factories) { this.factories = factories; }
    configure(name) { const Factory = this.factories.get(name); return new Factory().start(); }
  }
  return new ComponentConfigurator(new Map([["cache", Cache]])).configure("cache") === "cache:ready";
}

function interceptorContract() {
  const auditInterceptor = (request, nextHandler) => { request.audit = "recorded"; return nextHandler(request); };
  return auditInterceptor({ operation: "save" }, (request) => request.audit) === "recorded";
}

function extensionInterfaceContract() {
  class Diagnostics { status() { return "diagnostics:ready"; } }
  class CatalogueComponent { extension(name) { return name === "diagnostics" ? new Diagnostics() : undefined; } }
  return new CatalogueComponent().extension("diagnostics").status() === "diagnostics:ready";
}

function asynchronousCompletionTokenContract() {
  class CompletionToken { constructor(requestId) { this.requestId = requestId; this.result = undefined; } complete(result) { this.result = result; } }
  const token = new CompletionToken("run-42");
  token.complete("saved");
  return token.requestId === "run-42" && token.result === "saved";
}

function acceptorConnectorContract() {
  class Acceptor { accept(peer) { return `connected:${peer}`; } }
  class Connector { constructor(acceptor) { this.acceptor = acceptor; } connect(peer) { return this.acceptor.accept(peer); } }
  return new Connector(new Acceptor()).connect("catalogue-client") === "connected:catalogue-client";
}

function scopedLockingContract() {
  class ScopedLock {
    #locked = false;
    run(action) {
      if (this.#locked) throw new Error("The protected state is already locked.");
      this.#locked = true;
      try { return action(); } finally { this.#locked = false; }
    }
    get isLocked() { return this.#locked; }
  }
  const lock = new ScopedLock();
  const state = lock.run(() => "updated");
  return state === "updated" && !lock.isLocked;
}

function strategizedLockingContract() {
  class RecordingLock {
    constructor() { this.events = []; }
    acquire() { this.events.push("acquire"); }
    release() { this.events.push("release"); }
  }
  class ProtectedCounter {
    constructor(lockStrategy) { this.lockStrategy = lockStrategy; this.value = 0; }
    increment() {
      this.lockStrategy.acquire();
      try { this.value += 1; return this.value; } finally { this.lockStrategy.release(); }
    }
  }
  const lockStrategy = new RecordingLock();
  return new ProtectedCounter(lockStrategy).increment() === 1 && lockStrategy.events.join(",") === "acquire,release";
}

function threadSafeInterfaceContract() {
  class SafeInventory {
    #items = [];
    #busy = false;
    #synchronize(action) {
      if (this.#busy) throw new Error("Concurrent access is not allowed.");
      this.#busy = true;
      try { return action(); } finally { this.#busy = false; }
    }
    add(item) { this.#synchronize(() => this.#items.push(item)); }
    count() { return this.#synchronize(() => this.#items.length); }
  }
  const inventory = new SafeInventory();
  inventory.add("catalogue");
  return inventory.count() === 1;
}

function doubleCheckedLockingContract() {
  class LazyCatalogue {
    #instance;
    creations = 0;
    instance() {
      if (this.#instance === undefined) {
        if (this.#instance === undefined) {
          this.#instance = { status: "ready" };
          this.creations += 1;
        }
      }
      return this.#instance;
    }
  }
  const catalogue = new LazyCatalogue();
  return catalogue.instance() === catalogue.instance() && catalogue.creations === 1;
}

function threadSpecificStorageContract() {
  class RequestContexts {
    constructor() { this.contexts = new Map(); }
    setRequestId(threadId, requestId) { this.contexts.set(threadId, { requestId }); }
    requestIdFor(threadId) { return this.contexts.get(threadId)?.requestId; }
  }
  const contexts = new RequestContexts();
  contexts.setRequestId("worker-1", "run-42");
  return contexts.requestIdFor("worker-1") === "run-42" && contexts.requestIdFor("worker-2") === undefined;
}

function distributedTracingContract() {
  class Trace { constructor(traceId) { this.traceId = traceId; this.spans = []; } record(service) { this.spans.push(`${this.traceId}:${service}`); } }
  const trace = new Trace("trace-42");
  trace.record("catalogue-api"); trace.record("pricing");
  return trace.spans.join(",") === "trace-42:catalogue-api,trace-42:pricing";
}

function exceptionTrackingContract() {
  class ExceptionTracker { constructor() { this.reports = []; } record(service, error) { this.reports.push({ service, error: error.message }); } }
  const tracker = new ExceptionTracker();
  try { throw new Error("catalogue unavailable"); } catch (error) { tracker.record("catalogue-api", error); }
  return JSON.stringify(tracker.reports) === '[{"service":"catalogue-api","error":"catalogue unavailable"}]';
}

function parseCatalog(path) {
  return readFileSync(path, "utf8")
    .split(/\r?\n/)
    .filter((line) => line && !line.startsWith("#"))
    .map((line, index) => {
      const fields = line.split("|");
      if (fields.length !== 5) throw new Error(`Malformed manifest record on line ${index + 1}`);
      const [identifier, catalogues, family, name, contract] = fields;
      return { identifier, catalogues: catalogues.split(";"), family, name, contract };
    });
}

const contracts = Object.freeze({
  boundary: () => new Map([["request", "accepted"]]).get("request") === "accepted",
  blackboard: blackboardContract,
  broker: brokerContract,
  pac: pacContract,
  reflection: reflectionContract,
  "master-slave": masterSlaveContract,
  "command-processor": commandProcessorContract,
  "view-handler": viewHandlerContract,
  "forwarder-receiver": forwarderReceiverContract,
  "whole-part": wholePartContract,
  "client-dispatcher-server": clientDispatcherServerContract,
  "counted-pointer": countedPointerContract,
  "wrapper-facade": wrapperFacadeContract,
  "component-configurator": componentConfiguratorContract,
  interceptor: interceptorContract,
  "extension-interface": extensionInterfaceContract,
  "asynchronous-completion-token": asynchronousCompletionTokenContract,
  "acceptor-connector": acceptorConnectorContract,
  "scoped-locking": scopedLockingContract,
  "strategized-locking": strategizedLockingContract,
  "thread-safe-interface": threadSafeInterfaceContract,
  "double-checked-locking": doubleCheckedLockingContract,
  "thread-specific-storage": threadSpecificStorageContract,
  "distributed-tracing": distributedTracingContract,
  "exception-tracking": exceptionTrackingContract,
  composition: () => ["first", "second"].join("|") === "first|second",
  concurrency: () => new Set(["leader"]).size === 1,
  deployment: () => new Set(["region-a", "region-b"]).size === 2,
  mapping: () => ({ external: "internal" }).external === "internal",
  messaging: () => ((message) => message.id === "m-1")({ id: "m-1" }),
  observability: () => ["trace-1", "healthy"].every(Boolean),
  ordering: () => [3, 1, 2].sort((left, right) => left - right).join(",") === "1,2,3",
  persistence: () => new Map([["id", "saved"]]).get("id") === "saved",
  resilience: () => [1, 2, 3].find((attempt) => attempt === 2) === 2,
  routing: () => ("invoice".startsWith("invoice") ? "billing" : "support") === "billing",
  security: () => ({ token: "scoped" }).token === "scoped",
  state: () => "ready" === "ready",
});

function verify(definitions) {
  if (definitions.length !== EXPECTED_PATTERN_COUNT) {
    throw new Error(`Expected ${EXPECTED_PATTERN_COUNT} records, found ${definitions.length}`);
  }
  const identifiers = new Set(definitions.map((definition) => definition.identifier));
  if (identifiers.size !== definitions.length) throw new Error("Pattern identifiers must be unique");
  for (const definition of definitions) {
    if (![definition.identifier, definition.catalogues.length, definition.family, definition.name].every(Boolean)) {
      throw new Error(`Incomplete definition: ${definition.identifier}`);
    }
    const contract = contracts[definition.contract];
    if (!contract || !contract()) throw new Error(`${definition.name} contract failed`);
    console.log(`PASS ${definition.name}`);
  }
}

const manifest = join(__dirname, "..", "shared", "pattern-catalog.tsv");
verify(parseCatalog(manifest));
console.log(`Verified ${EXPECTED_PATTERN_COUNT} catalogued patterns.`);
